src/lib/openssl_preload/shadow_openssl_rng.c

```diff
@@ -0,0 +1,115 @@
+/*
+ * The Shadow Simulator
+ * Copyright (c) 2010-2011, Rob Jansen
+ * See LICENSE for licensing information
+ */
+
+/* Implements an LD_PRELOAD library intended for use with Shadow. libcrypto
+ * otherwise internally uses some entropy sources that Shadow is unable to trap
+ * and emulate (such as the RDRAND instruction), making Shadow simulations of
+ * software using libcrypto non-deterministic.
+ *
+ * To use this library, set LD_PRELOAD in the target program's `environment`
+ * attribute in the Shadow simulation config file. e.g.:
+ *
+ * hosts:
+ *   torclient:
+ *     processes:
+ *       - path: ~/.local/bin/tor
+ *         environment: 'LD_PRELOAD=~/.local/lib/libshadow_openssl_rng.so'
+ */
+
+#include <stddef.h>
+#include <sys/syscall.h>
+#include <unistd.h>
+
+static int _getRandomBytes(unsigned char* buf, int numBytes) {
+    // shadow interposes this and will fill the buffer for us
+    // return 1 on success, 0 otherwise
+    return (numBytes == syscall(SYS_getrandom, buf, (size_t)numBytes, 0)) ? 1 : 0;
+}
+
+int RAND_DRBG_generate(void *drbg,
+                        unsigned char *out, size_t outlen,
+                        int prediction_resistance,
+                        const unsigned char *adin, size_t adinlen) {
+    _getRandomBytes(out, outlen);
+    return 1;
+}
+
+int RAND_DRBG_bytes(void *drbg,
+                     unsigned char *out, size_t outlen) {
+    _getRandomBytes(out, outlen);
+    return 1;
+}
+
+int RAND_bytes(unsigned char *buf, int num) {
+    _getRandomBytes(buf, num);
+    return 1;
+}
+
+int RAND_pseudo_bytes(unsigned char *buf, int num) {
+    return _getRandomBytes(buf, num);
+}
+
+void RAND_seed(const void *buf, int num) {
+    return;
+}
+
+void RAND_add(const void *buf, int num, double entropy) {
+    return;
+}
+
+int RAND_poll() {
+    return 1;
+}
+
+// Some versions of openssl define RAND_cleanup as a macro.
+// Get that definition out of the way so we can override the library symbol in
+// case it still exists.
+#ifdef RAND_cleanup
+#undef RAND_cleanup
+#endif
```

Comment:
This shouldn't be needed since this C file doesn't import any openssl headers.

Reply:
Done